controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,136 @@
+package nstemplateset
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+)
+
+type spaceRolesManager struct {
+	*statusManager
+}
+
+// ensure ensures that the space roles for the users exist.
+// Returns `true, nil` if something was changed, `false, nil` if nothing changed, `false, err` if an error occurred
+// TODO: support update/removal/outdated
+func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	logger = logger.WithValues("nstemplateset_name", nsTmplSet.Name)
+
+	nss, err := fetchNamespacesByWorkspace(r.Client, nsTmplSet.Name)
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles to apply now
+		lastAppliedSpaceRoleObjs, err := r.getLastAppliedSpaceRoles(&ns)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
```

Comment:
what about checking the annotation for each of the namespace - if it matches, then we wouldn't do anything here, if it doesn't match, then it would process the templates and deleted/update/create objects that are needed. 
It would significantly decrease the load that is caused by processing the templates.

Reply:
good idea! done in 0034497